auth/couchdb_session_authenticator_test.go

```diff
@@ -0,0 +1,281 @@
+			err     error
+		)
+
+		BeforeEach(func() {
+			callNumber := 0
+			server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				callNumber++
+				cookie := &http.Cookie{
+					Name:  "AuthSession",
+					Value: fmt.Sprintf("fakefake-%d", callNumber),
+					Expires: time.Now().
+						Add(24*time.Hour - time.Duration(callNumber)*time.Minute),
+				}
+				http.SetCookie(w, cookie)
+				w.WriteHeader(http.StatusOK)
+			}))
+
+			builder, err := core.NewRequestBuilder(core.GET).
+				ResolveRequestURL(server.URL, "/db", nil)
+			Expect(err).To(BeNil())
+
+			ctx := context.WithValue(context.Background(), contextKey("key"), "abcdefgh")
+			request, err = builder.
+				AddHeader("X-Req-Id", "abcdefgh").
+				WithContext(ctx).
+				Build()
+			Expect(err).To(BeNil())
+			Expect(request).ToNot(BeNil())
+
+			auth, err = NewCouchDbSessionAuthenticator("user", "pass")
+			Expect(err).To(BeNil())
+			Expect(auth).ToNot(BeNil())
+			Expect(auth.AuthenticationType()).To(Equal(AUTHTYPE_COUCHDB_SESSION))
+		})
+
+		AfterEach(func() {
+			server.Close()
+		})
+
+		It("Test setting URL, Headers and Context on Authenticator", func() {
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			Expect(auth.url).To(Equal(server.URL))
+			Expect(auth.header).To(HaveKeyWithValue("X-Req-Id", []string{"abcdefgh"}))
+			Expect(auth.ctx.Value(contextKey("key"))).To(Equal("abcdefgh"))
+		})
+
+		It("Test setting custom http client on Authenticator", func() {
+			// set http.Client with custom timeout
+			auth.Client = &http.Client{Timeout: time.Second}
+
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			Expect(auth.Client.Timeout).To(Equal(time.Second))
+		})
+
+		It("Test authentication re-request after expiration", func(done Done) {
+			err = auth.Authenticate(request)
+			Expect(err).To(BeNil())
+
+			// Fetch a cookie from the cache to confirm it's present
+			// and prepare cache state for the next test
+			cookie, err := auth.getCookie()
+			Expect(err).To(BeNil())
+			Expect(cookie).ToNot(BeNil())
+			Expect(cookie.Name).To(Equal("AuthSession"))
+			Expect(cookie.Value).To(Equal("fakefake-1"))
+			Expect(cookie.Expires).ToNot(BeNil())
+
+			// Fetch cookie again to verify cache is working as expected,
+			// we are getting old cookie and refresh process's not triggered
+			oldRefresh := auth.session.refreshTime
+			cookie, err = auth.getCookie()
+			Expect(err).To(BeNil())
+			Expect(cookie.Value).To(Equal("fakefake-1"))
+			Expect(auth.session.refreshTime).To(Equal(oldRefresh))
+
+			// Force expiration and verify we got a new cookie
+			auth.session.expires = time.Now().Add(-time.Minute)
+
+			// Run getCookie in three parallel threads, to verify
+			// that request mutex works and we are querying /_session only once
+			var wg sync.WaitGroup
+
+			for i := 1; i <= 3; i++ {
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					cookie, err = auth.getCookie()
```

Comment:
After this line you should add verify the test creates only 1 new cookie.

Reply:
I don't understand what you are saying here, this is exactly what next three test are verifying. Please elaborate.